pkg/controller/function/function_controller.go

```diff
@@ -83,62 +107,178 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		return reconcile.Result{}, err
 	}
 
-	if len(fnConfigNamespaceEnv) > 0 {
-		fnConfigNamespace = fnConfigNamespaceEnv
+	// Get Function instance
+	fn := &runtimev1alpha1.Function{}
+	if err := r.getFunctionInstance(request, fn); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error reading Function instance", "namespace", request.Namespace, "name", request.Name)
+		return reconcile.Result{}, err
 	}
-	fnConfig := &corev1.ConfigMap{}
-	err := r.Get(context.TODO(), types.NamespacedName{Name: fnConfigName, Namespace: fnConfigNamespace}, fnConfig)
 
-	if err != nil {
-		log.Error(err, "Unable to read Function controller config: %v from Namespace: %v", fnConfigName, fnConfigNamespace)
+	// Create Function's ConfigMap
+	foundCm := &corev1.ConfigMap{}
+	deployCm := &corev1.ConfigMap{}
+	if err := r.createFunctionConfigMap(foundCm, deployCm, fn); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error while trying to create the Function's ConfigMap", "namespace", deployCm.Namespace, "name", deployCm.Name)
 		return reconcile.Result{}, err
 	}
 
-	rnInfo, err := runtimeUtil.New(fnConfig)
-	if err != nil {
-		fmt.Printf("Error in reading ConfigMap: %v", err)
+	// Update Function's ConfigMap
+	if err := r.updateFunctionConfigMap(foundCm, deployCm); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		log.Error(err, "Error while trying to update Function's ConfigMap:", "namespace", deployCm.Namespace, "name", deployCm.Name)
 		return reconcile.Result{}, err
 	}
-	// Fetch the Function instance
-	fn := &runtimev1alpha1.Function{}
-	err = r.Get(context.TODO(), request.NamespacedName, fn)
+
+	// Create function's image name
+	hash := sha256.New()
+	print("foundCM: %v", foundCm)
+	hash.Write([]byte(foundCm.Data["handler.js"] + foundCm.Data["package.json"]))
+	functionSha := fmt.Sprintf("%x", hash.Sum(nil))
+	imageName := fmt.Sprintf("%s/%s-%s:%s", rnInfo.RegistryInfo, fn.Namespace, fn.Name, functionSha)
+	log.Info("function image", "namespace:", fn.Namespace, "name:", fn.Name, "imageName:", imageName)
+
+	if err := r.getFunctionBuildTemplate(rnInfo, fnConfig, fn); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		return reconcile.Result{}, err
+	}
+	log.Info("get BuildTemplate [done]", "function_name:", fn.Name)
+
+	if err := r.buildFunctionImage(rnInfo, fnConfig, fn, imageName); err != nil {
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
+
+		return reconcile.Result{}, err
+	}
+	log.Info("build Function image [done]", "function_name:", fn.Name)
```

Comment:
We should remove all logs with the pattern `".* [done]"`. It was meant for debugging only and bloates the log output probably to much.

Reply:
Yes, it is not needed. If there is an error this will happens when the Build is being created.